Fix #3298, #3299 and #3301: Introduce UI for MarkChaptersCompleted, MarkStoriesCompleted and MarkTopicsCompleted

yashraj-01 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fixes #3298, #3299 and #3301: Introduced UI for MarkChaptersCompleted, MarkStoriesCompleted and MarkTopicsCompleted.

**Note:** 

- This PR only introduces the UI and the logic for displaying the list of chapters/stories/topics. All the interaction and modifying progress logic will be tracked by issues #3303, #3305 and #3306.

- In this PR I have added all the newly added kotlin files to `test_file_exemptions.textproto` except:

1. `DeveloperOptionsActivity`
2. `MarkChaptersCompletedActivity`
3. `MarkStoriesCompletedActivity`
4. `MarkTopicsCompletedActivity`
5. `ViewEventLogsActivity`

- This is to prevent CI checks from failing.

**Mock links:** 

| MarkChaptersCompleted | MarkStoriesCompleted | MarkTopicsCompleted |
|:---:|:---:|:---:|
| https://xd.adobe.com/view/e8aa4198-3940-47f9-514a-f41cc54457f6-9e9b/screen/ae50eed5-c5c5-4fc4-b5a8-fef1748f782a/ | https://xd.adobe.com/view/e8aa4198-3940-47f9-514a-f41cc54457f6-9e9b/screen/8a5fe55c-6244-4702-970d-97eea6f3bb2a/ | https://xd.adobe.com/view/e8aa4198-3940-47f9-514a-f41cc54457f6-9e9b/screen/db9d6aaf-5588-4963-8f32-1351677bcf47/ |

**Screenshots**
| MarkChaptersCompleted | MarkStoriesCompleted | MarkTopicsCompleted |
|:---:|:---:|:---:|
| <img width="243" height="432" src="https://user-images.githubusercontent.com/63469151/122816554-83cc7580-d2f4-11eb-8522-208fc183179d.gif">  | <img width="243" height="432" src="https://user-images.githubusercontent.com/63469151/122816661-a52d6180-d2f4-11eb-9c5a-ef4b2cb7a038.gif"> | <img width="243" height="432" src="https://user-images.githubusercontent.com/63469151/122816725-bbd3b880-d2f4-11eb-804d-659b011bd23f.gif"> |

- `DeveloperOptionsActivityTest`

| Espresso | Robolectric |
|:---:|:---:|
| ![Screenshot from 2021-07-07 00-11-42](https://user-images.githubusercontent.com/63469151/124651178-3c2f1780-deb8-11eb-811d-2348ee656f78.png) | ![Screenshot from 2021-07-07 00-12-57](https://user-images.githubusercontent.com/63469151/124651294-5b2da980-deb8-11eb-9f9f-67270236bed9.png) |

- `MarkChaptersCompletedActivityTest`

| Espresso | Robolectric |
|:---:|:---:|
| ![Screenshot from 2021-07-06 02-32-45](https://user-images.githubusercontent.com/63469151/124522690-c3758000-de11-11eb-8be1-a3d52c1d2597.png) | ![Screenshot from 2021-07-06 03-50-05](https://user-images.githubusercontent.com/63469151/124522709-d5572300-de11-11eb-949a-d35214717c17.png) |

- `MarkStoriesCompletedActivityTest`

| Espresso | Robolectric |
|:---:|:---:|
| ![Screenshot from 2021-07-07 00-19-02](https://user-images.githubusercontent.com/63469151/124651865-15251580-deb9-11eb-92c2-167acedefb5b.png) | ![Screenshot from 2021-07-07 00-19-34](https://user-images.githubusercontent.com/63469151/124651892-1e15e700-deb9-11eb-957b-b9737302397a.png) |

- `MarkTopicsCompletedActivityTest`

| Espresso | Robolectric |
|:---:|:---:|
| ![Screenshot from 2021-07-07 00-25-43](https://user-images.githubusercontent.com/63469151/124652786-23276600-deba-11eb-8e9f-b65f59e409e7.png) | ![Screenshot from 2021-07-07 00-26-15](https://user-images.githubusercontent.com/63469151/124652816-2ae70a80-deba-11eb-95c8-f29e24042999.png) |

- Robolectric test results for `ModifyLessonProgressControllerTest`
![Screenshot from 2021-07-05 22-50-48](https://user-images.githubusercontent.com/63469151/124503818-ec7f1c00-dde3-11eb-9cdd-001714858fc3.png)

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

yashraj-01 committed:
Implement Developer Options Starter

yashraj-01 committed:
Fix lint issues

yashraj-01 committed:
Implement Developer Options Menu Item in Nav Drawer

yashraj-01 committed:
Fix lint issues

yashraj-01 committed:
Merge branch 'develop' into developer-options-starter

yashraj-01 committed:
Added Kdoc. abstract class -> interface

yashraj-01 committed:
Merge branch 'developer-options-starter' into developer-options-menu-item

yashraj-01 committed:
Added tests for Developer Options menu item

yashraj-01 committed:
Fixed lint issues

yashraj-01 committed:
Merge branch 'develop' into developer-options-starter

yashraj-01 committed:
Introduced ApplicationComponentForDevMode

yashraj-01 committed:
Merge branch 'developer-options-starter' into developer-options-menu-item

yashraj-01 committed:
Added KDocs and fixed lint issues

yashraj-01 committed:
Merge branch 'developer-options-starter' into developer-options-menu-item

yashraj-01 committed:
Added test to check unavailibility of Dev Options menu item when not in dev mode

yashraj-01 committed:
Added initial test for DeveloperOptionsActivity. Added KDocs. Fixed lint issues

yashraj-01 committed:
Introduced DebugApplicationComponent

yashraj-01 committed:
Merge branch 'developer-options-starter' into developer-options-menu-item

yashraj-01 committed:
Implement developer options menu ui

yashraj-01 committed:
Modified way of passing data to recycler view

yashraj-01 committed:
removed unnecessary isMultipane boolean

yashraj-01 committed:
added view models to app/BUILD.bazel

yashraj-01 committed:
Added tests for DeveloperOptionsActivityTest

yashraj-01 committed:
Fixed lint issues

yashraj-01 committed:
Merge branch 'develop' into developer-options-starter

yashraj-01 committed:
Merge branch 'developer-options-starter' into developer-options-menu-item

yashraj-01 committed:
Merge branch 'developer-options-menu-item' into developer-options-menu

yashraj-01 committed:
Corrected key mismatch error. Changed isDevMode to isDebugMode. Nit fixes

yashraj-01 committed:
Fixed Robolectric test for NavigationDrawerActivityTest

yashraj-01 committed:
Updated developer_options_fragment layout

yashraj-01 committed:
Resolved merge conflicts

yashraj-01 committed:
Merge branch 'develop' into developer-options-starter

yashraj-01 committed:
Resolved merge conflicts

yashraj-01 committed:
Merge branch 'developer-options-menu-item' into developer-options-menu

yashraj-01 committed:
Added test for hints switch

yashraj-01 committed:
Added eof. Changed dev mode to debug mode

yashraj-01 committed:
resolved merge conflicts

yashraj-01 committed:
Initial progress on UI of MarkChaptersCompleted

yashraj-01 committed:
Fixed lint issues

yashraj-01 committed:
Fixed populating recyclerview

yashraj-01 committed:
Fixed lint issues

yashraj-01 committed:
Removed unnecessary lastLoadedFragment flag

yashraj-01 committed:
Resolved merge conflicts and removed unnecessary code

yashraj-01 committed:
fixed lint issues

yashraj-01 committed:
Added dependencies to app/BUILD.bazel

yashraj-01 committed:
Corrected imports

yashraj-01 committed:
Removed unnecessary code from DeveloperOptionsActivityTest

yashraj-01 committed:
Resolved merge conflicts

yashraj-01 committed:
Resolved merge conflicts

yashraj-01 committed:
Modified approach for Application Components

yashraj-01 committed:
nit fix

yashraj-01 committed:
nit fix

yashraj-01 committed:
Implemented UI of MarkStoriesCompleted

yashraj-01 committed:
Implemented UI of MarkTopicsCompleted

yashraj-01 committed:
Added tests and Kdocs + nit fixes

yashraj-01 committed:
Merge branch 'develop' into developer-options-starter

yashraj-01 committed:
Resolved merge conflicts

yashraj-01 committed:
Code reformating

yashraj-01 committed:
Separated dev and prod tests for NavigationDrawerActivityTest

yashraj-01 committed:
Resolved merge conflicts

yashraj-01 committed:
Resolved merge conflicts

yashraj-01 committed:
Added tests to check intents in DeveloperOptionsActivityTest

yashraj-01 committed:
Added tests for MarkChaptersCompletedActivityTest

yashraj-01 committed:
Added tests for MarkStoriesCompletedActivityTest

yashraj-01 committed:
Add tests for MarkTopicsCompletedActivityTest

yashraj-01 commented:
**References:**

- Comment link: https://docs.google.com/document/d/1DgVxNsHxgEhfx35iKdCcEAMQSSd0jqLqq9H2F0pjS1w/edit?disco=AAAAM4NaoaU
- Comment screenshot (to save time)
![image](https://user-images.githubusercontent.com/63469151/123323394-11a59c00-d553-11eb-9b8b-e0c429474efa.png)
- **Gist link:** https://gist.github.com/yashraj-01/3cc650f459fe5689b18ea394df3a1cdd

Currently, I have implemented Method 1 in this PR.

@BenHenning @rt4914 @Sarthak2601 I request you to please take a look at the current implementation and the gist that I wrote and provide your suggestions on which approach we should follow. If you think there could be a better approach then please let me know. Thanks.

**Please Note:** The logic to modify lesson progress is not the scope of this PR. This PR only corresponds to displaying the list of chapters/stories/topics.

yashraj-01 committed:
nit fix

fsharpasharp reviewed (approved):
Approving for BUILD.bazel changes.

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/devoptions/DeveloperOptionsActivityTest.kt:
```diff
@@ -241,6 +249,48 @@ class DeveloperOptionsActivityTest {
+  fun testDeveloperOptionsFragment_clickMarkChaptersCompleted_opensMarkChaptersCompletedActivity() {
+    launch<DeveloperOptionsActivity>(
+      createDeveloperOptionsActivityIntent(
+        internalProfileId = internalProfileId
```
we can skip such named arguments where the named argument is the same as the argument.

yashraj-01 replied:
removed at all occurrences.

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt:
```diff
@@ -0,0 +1,462 @@
+      testCoroutineDispatchers.runCurrent()
+      scrollToPosition(0)
+      onView(
+        atPositionOnView(
```
a helper function in this file is helpful same as we have in other test files. 
Refer from - `DeveloperOptionsActivityTest`

yashraj-01 replied:
done

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt:
```diff
@@ -0,0 +1,462 @@
+  }
+
+  @Test
+  fun testMarkChaptersCompletedActivity_configChange_chapterListSAreVisible() {
```
```suggestion
  fun testMarkChaptersCompletedActivity_configChange_chapterListIsVisible() {
```

yashraj-01 replied:
done

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt:
```diff
@@ -0,0 +1,462 @@
+  }
+
+  @Test
+  fun testMarkChaptersCompletedActivity_selectAll_thenDeselect_isNotChecked() {
```
Found something interesting about `Select-Deselect`. 

Google Defination
```
Deselect is the real antonym of select. Unselect is only used when talking about computers. 
```

Also, no need of `then`

yashraj-01 replied:
replaced `deselect` with `unselect`. also removed `then`

anandwana001 commented on app/src/main/res/values/strings.xml:
```diff
@@ -64,6 +64,9 @@
   <string name="practice">Practice</string>
   <string name="revision">Revision</string>
   <string name="developer_options_activity_title">Developer Options</string>
+  <string name="mark_chapters_completed_activity_title">Mark Chapters Completed</string>
```
though this file doesn't come under my codeowners, nit suggestion, It will be helpful to keep these new strings under a section just like we have a section for `  <!-- DeveloperOptionsFragment -->`

yashraj-01 replied:
I think we should create a separate issue to put all the activity titles under one section such as `<!-- Activity Titles -->`. WDYT?

anandwana001 replied:
Sounds good to me, let's create a good first issue for any Home activity only. I will create a good first issue for other screens later on. 

For Developer Options, suggesting to fix in this PR only.

yashraj-01 replied:
I have arranged the titles. PTAL. Also, I have filed an issue for the same (#3384).

anandwana001 replied:
Earlier I got the wrong idea from activity Titels, I think it will be replaced with the activity name. 

Suggest in review, take a look at comments below.

yashraj-01 replied:
Apologies for the delay. I have addressed all your below comments. PTAL. Thanks.

anandwana001 reviewed (changes requested):
Added nit suggestions, assign me back for full review.

yashraj-01 committed:
Renamed NavigationDrawerActivityDevTest to NavigationDrawerActivityDebugTest

yashraj-01 committed:
nit fixes. created helper functions

rt4914 reviewed (changes requested):
@yashraj-01 Will review it once initial few PRs are merged as a minor change in those PRs will propagate to this PR too.

yashraj-01 committed:
Moved starter to devoptions

yashraj-01 committed:
Moved starter to devoptions

yashraj-01 committed:
Merge branch 'developer-options-starter' into developer-options-menu-item

yashraj-01 committed:
Resolved merge conflicts

vinitamurthi reviewed (approved):
LGTM for codeowner files!

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt:
```diff
@@ -0,0 +1,380 @@
+      createMarkChaptersCompletedActivityIntent(internalProfileId)
+    ).use {
+      testCoroutineDispatchers.runCurrent()
+      scrollToPosition(0)
```
add named arguments

yashraj-01 replied:
Done.

anandwana001 reviewed (changes requested):
Nit things, else LGTM

yashraj-01 committed:
nit fix

BenHenning commented:
Similar to earlier PRs, I'm not sure what I should review here. Please clarify @yashraj-01 / @Sarthak2601.

yashraj-01 committed:
Resolved merge conflicts

yashraj-01 committed:
Added TODOs

yashraj-01 committed:
Added named arguments

yashraj-01 committed:
Arranged activity titles in app/strings.xml

yashraj-01 commented:
> Similar to earlier PRs, I'm not sure what I should review here. Please clarify @yashraj-01 / @Sarthak2601.

Apologies for the inconvenience. Will update the reviewers and assignee sections.

yashraj-01 commented:
Apologies, @vinitamurthi. I accidentally clicked on Re-request review.

yashraj-01 committed:
Removed unnecessary code. Nit fixes

yashraj-01 committed:
Merge branch 'develop' into developer-options-menu-item

yashraj-01 committed:
Merge branch 'developer-options-menu-item' into developer-options-menu

anandwana001 commented on app/src/main/res/values/strings.xml:
```diff
@@ -186,6 +187,11 @@
     </item>
   </plurals>
   <string name="bar_separator">\u0020|\u0020</string>
+  <!-- Activity Titles -->
```
This should be name as file, look for other for reference.

yashraj-01 replied:
Renamed to `<!-- DeveloperOptionsActivity -->`.

anandwana001 commented on app/src/main/res/values/strings.xml:
```diff
@@ -105,6 +104,8 @@
   <string name="profile_progress_activity_title">Profile Progress Page</string>
   <string name="search">Search</string>
+  <string name="all">All</string>
+  <string name="mark_completed_uppercase">MARK COMPLETED</string>
```
Comes under developer options section

yashraj-01 replied:
Done.

yashraj-01 committed:
Resolved merge conflicts

yashraj-01 committed:
Added accessiblity test rule. Modified UI for the same

yashraj-01 committed:
Nit fix

yashraj-01 committed:
Created ModifyLessonProgressController

yashraj-01 committed:
Added TODO

yashraj-01 committed:
Added Kdoc to DeveloperOptionsViewModel

yashraj-01 committed:
Added KDocs

yashraj-01 committed:
Added method to getAllStoriesWithProgress

yashraj-01 committed:
Modified approach to get chapter list with progress

yashraj-01 committed:
Added tests related to all stories in ModifyLessonProgressControllerTest

yashraj-01 committed:
Added check if completed functions and tests for the same

yashraj-01 committed:
Added isCompleted flag to topi and story view models

yashraj-01 committed:
Merge branch 'develop' into developer-options-menu

yashraj-01 committed:
Nit fixes

yashraj-01 committed:
Merge branch 'develop' into developer-options-menu

yashraj-01 committed:
Nit changes in strings.xml

yashraj-01 committed:
Resolved merge conflicts

yashraj-01 committed:
Added KDocs + nit fixes

yashraj-01 commented:
> **References:**
> 
>     * Comment link: https://docs.google.com/document/d/1DgVxNsHxgEhfx35iKdCcEAMQSSd0jqLqq9H2F0pjS1w/edit?disco=AAAAM4NaoaU
> 
>     * Comment screenshot (to save time)
>       ![image](https://user-images.githubusercontent.com/63469151/123323394-11a59c00-d553-11eb-9b8b-e0c429474efa.png)
> 
>     * **Gist link:** https://gist.github.com/yashraj-01/3cc650f459fe5689b18ea394df3a1cdd
> 
> 
> Currently, I have implemented Method 1 in this PR.
> 
> @BenHenning @rt4914 @Sarthak2601 I request you to please take a look at the current implementation and the gist that I wrote and provide your suggestions on which approach we should follow. If you think there could be a better approach then please let me know. Thanks.
> 
> **Please Note:** The logic to modify lesson progress is not the scope of this PR. This PR only corresponds to displaying the list of chapters/stories/topics.

As per discussion with @rt4914 I have created `ModifyLessonProgressController` and put all the logic in the domain layer. It has also significantly changed the app layer files. So, I request mentors to please take a fresh pass. Thanks.

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/devoptions/DeveloperOptionsActivityTest.kt:
```diff
@@ -237,6 +245,42 @@ class DeveloperOptionsActivityTest {
   }
 
+  @Test
+  fun testDeveloperOptionsFragment_clickMarkChaptersCompleted_opensMarkChaptersCompletedActivity() {
```
can we have `configChange` variant of these 3 new test cases.

yashraj-01 replied:
Tests added.

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt:
```diff
@@ -0,0 +1,380 @@
+  }
+
+  @Test
+  fun testMarkChaptersCompletedActivity_selectAll_isChecked() {
```
add 1 more test `testMarkChaptersCompletedActivity_selectAll_configCahnge_isChecked`

yashraj-01 replied:
Done.

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt:
```diff
@@ -0,0 +1,380 @@
+  }
+
+  @Test
+  fun testMarkChaptersCompletedActivity_selectAll_unselect_isNotChecked() {
```
add 1 more test `testMarkChaptersCompletedActivity_selectAll_unselect_configChange_isNotChecked`

yashraj-01 replied:
This test is now redundant. I have removed it. Will add new tests with their configChange variants.

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt:
```diff
@@ -0,0 +1,380 @@
+  }
+
+  @Test
+  fun testMarkChaptersCompletedActivity_selectChaptersOfFirstStory_chaptersAreChecked() {
```
add 1 more test `testMarkChaptersCompletedActivity_selectChaptersOfFirstStory_configChange_chaptersAreChecked`

yashraj-01 replied:
Done.

anandwana001 reviewed (changes requested):
Left few suggestions, apply in other newly added test files also.

yashraj-01 committed:
Corrected imports in test files

Sarthak2601 commented on app/src/main/java/org/oppia/android/app/devoptions/devoptionsitemviewmodel/DeveloperOptionsModifyLessonProgressViewModel.kt:
```diff
@@ -1,7 +1,31 @@
+  private val routeToMarkTopicsCompletedListener: RouteToMarkTopicsCompletedListener
+) : DeveloperOptionsItemViewModel() {
+
+  /** Called when the 'mark chapters completed' button is clicked by the user. */
```
I think `called when button is clicked` can be replaced by `routes the user to the <screen_name> screen for <use-case>. ` Ditto for others as well.

yashraj-01 replied:
KDoc updated. PTAL.

Sarthak2601 commented on app/src/main/java/org/oppia/android/app/devoptions/markchapterscompleted/MarkChaptersCompletedFragment.kt:
```diff
@@ -0,0 +1,51 @@
+  }
+
+  @Inject
+  lateinit var markChaptersCompletedFragmentPresenter: MarkChaptersCompletedFragmentPresenter
```
Move this above the companion object.

yashraj-01 replied:
Done. PTAL.

Sarthak2601 commented on domain/src/main/java/org/oppia/android/domain/topic/ModifyLessonProgressController.kt:
```diff
@@ -0,0 +1,109 @@
+package org.oppia.android.domain.topic
```
This file has to go in a separate devoptions package in domain module.

yashraj-01 replied:
![Screenshot from 2021-07-05 22-51-32](https://user-images.githubusercontent.com/63469151/124503644-9b6f2800-dde3-11eb-828e-fead4e12da59.png)
Actually, @BenHenning mentioned to use this `devoptions` package specifically for the UI part only. That's why I put this controller in the `topic` package. Should I move it to `devoptions` package in `domain` module? @Sarthak2601

yashraj-01 replied:
Moved to `devoptions` package. PTAL.

Sarthak2601 reviewed (changes requested):
Great work @yashraj-01! Suggested some changes. Additionally as Akshay mentioned in a few files, you'll have to write config change tests in almost all your app-module tests.

yashraj-01 committed:
Fixed cobine topi and topicProgress function

yashraj-01 committed:
Implemented topic selection logic

yashraj-01 committed:
Implemented story selection logic

yashraj-01 committed:
Nit fixes

vinitamurthi commented:
Hi @yashraj-01 , this seems to be mainly UI work. Is there anything I need to take a pass on?

yashraj-01 committed:
Added disabled checkbox color

yashraj-01 committed:
Changed controller method to return map of stories and topic id. Changed from nested recycler view to MultiType recycler view in MarkChaptersCompleted

yashraj-01 committed:
Fixed recycler view crashing

yashraj-01 committed:
Implemented all check for MarkChaptersCompleted

yashraj-01 committed:
Corrected tests in ModifyLessonProgressControllerTest

yashraj-01 commented:
> Hi @yashraj-01 , this seems to be mainly UI work. Is there anything I need to take a pass on?

Hi @vinitamurthi , I think you could take a pass on the domain files. Thanks.

yashraj-01 committed:
Nit fixes

yashraj-01 committed:
Added configChange tests in DeveloperOptionsActivityTest. Corrected tests in MarkChapters/Stories/TopicsCompletedActivityTest

yashraj-01 committed:
Added more tests to MarkChaptersCompletedActivityTest

yashraj-01 committed:
Changed UI according to accessibilityTestRule

yashraj-01 committed:
Modified UI according to accessibilityTestRule for stories and topics as well

yashraj-01 committed:
Moved ModifyLessonProgressController to devoptions package

BenHenning reviewed (commented):
@yashraj-01 now that #3331 is merged can you update this with the latest changes from develop & then assign back? It seems that some of the deltas are no longer applicable which makes it difficult to see the new Bazel uses.

yashraj-01 committed:
Resolved merge conflicts

yashraj-01 committed:
Nits and kdocs

yashraj-01 committed:
Fixed bazel build failure

yashraj-01 committed:
Added more tests for MarkStoriesCompleted

yashraj-01 committed:
Added more tests for MarkTopicsCompleted

yashraj-01 committed:
Added TODOs and nit fixes

yashraj-01 committed:
Modifed chapters recycler view interaction logic

yashraj-01 commented:
> Great work @yashraj-01! Suggested some changes. Additionally as Akshay mentioned in a few files, you'll have to write config change tests in almost all your app-module tests.

@anandwana001 @Sarthak2601 I have added all possible tests that I could think of. PTAL. Also, I have addressed all your comments. PTAL. Thanks.

yashraj-01 commented:
> @yashraj-01 now that #3331 is merged can you update this with the latest changes from develop & then assign back? It seems that some of the deltas are no longer applicable which makes it difficult to see the new Bazel uses.

I have updated the PR. All merge conflicts are resolved now. Please take a pass. Thanks.

yashraj-01 commented:
@rt4914 could you please take a pass on the app layer files? Thanks.

anandwana001 commented:
@yashraj-01 Assign me back once it is approved from the UI layer, so we know that the only thing left is a test to review. Preventing a possibility where test files get approved but there might some change in UI, in that case, we need to update the test too and re-review.

BenHenning commented on app/src/main/java/org/oppia/android/app/devoptions/markchapterscompleted/ChapterSummaryViewModel.kt:
```diff
@@ -0,0 +1,20 @@
+  val storyId: String,
+  val topicId: String
+) : MarkChaptersCompletedItemViewModel() {
+  /**
+   * Function to check if a chapter is completed or not. It is used in
+   * [MarkChaptersCompletedFragmentPresenter] to decide if a chapter should be checked and disabled.
+   */
```
Function KDocs should always read as a verb phrase. Consider this alternative:

```kotlin
/** Returns whether the chapter represented by the current view model is completed. */
```

Note the phrasing. Also, I omitted the "It is used" bit since that's unrelated to this class. We should only ever document what a function does and how it can be used, not how it's actually used. The latter is always changing & is better found by analyzing call sites rather than documentation.

Ditto elsewhere for all other KDocs.

yashraj-01 replied:
Thanks for the suggestion. I have updated the Kdoc.

BenHenning commented on app/src/main/java/org/oppia/android/app/devoptions/markchapterscompleted/MarkChaptersCompletedItemViewModel.kt:
```diff
@@ -0,0 +1,5 @@
+
+import org.oppia.android.app.viewmodel.ObservableViewModel
+
+abstract class MarkChaptersCompletedItemViewModel : ObservableViewModel()
```
Please add missing KDoc.

Ditto elsewhere (if applicable).

yashraj-01 replied:
KDoc added.

BenHenning commented on app/src/main/java/org/oppia/android/app/devoptions/markchapterscompleted/MarkChaptersCompletedViewModel.kt:
```diff
@@ -0,0 +1,98 @@
+  private val modifyLessonProgressController: ModifyLessonProgressController
+) : ObservableViewModel() {
+
+  private var internalProfileId: Int = -1
```
We ought to be passing around a ``ProfileId`` instead of the raw integer. I suggest updating this accordingly.

yashraj-01 replied:
Done.

BenHenning commented on app/src/main/java/org/oppia/android/app/devoptions/markstoriescompleted/StorySummaryViewModel.kt:
```diff
@@ -0,0 +1,12 @@
+import org.oppia.android.app.model.StorySummary
+import org.oppia.android.app.viewmodel.ObservableViewModel
+
+/** [ViewModel] for displaying a story summary. */
```
This documentation is a bit vague--can you add more context as to exactly how this view model should be used & what it represents?

yashraj-01 replied:
KDoc updated. PTAL.

BenHenning commented on app/src/main/java/org/oppia/android/app/devoptions/markstoriescompleted/MarkStoriesCompletedViewModel.kt:
```diff
@@ -0,0 +1,83 @@
+  private val modifyLessonProgressController: ModifyLessonProgressController
+) : ObservableViewModel() {
+
+  private var internalProfileId: Int = -1
```
Ditto here.

yashraj-01 replied:
Done.

BenHenning commented on app/src/main/java/org/oppia/android/app/devoptions/marktopicscompleted/MarkTopicsCompletedViewModel.kt:
```diff
@@ -0,0 +1,78 @@
+   * List of topic ids used in [MarkTopicsCompletedFragmentPresenter] to check if all topics are
+   * selected or not.
+   */
+  val availableTopicIdList = ArrayList<String>()
```
More Kotlin idiomatic would be to create a mutable list using ``mutableListOf<String>()``, however it's a bit concerning to expose a mutable val as part of a class's public API. Is this list directly modified by databinding? If not, we should add methods to change the list rather than exposing the list itself.

yashraj-01 replied:
No, it is not directly modified by the binding. I have made this private and created a getter function for it. PTAL.

BenHenning commented on app/src/main/java/org/oppia/android/app/devoptions/marktopicscompleted/TopicSummaryViewModel.kt:
```diff
@@ -0,0 +1,8 @@
+import org.oppia.android.app.model.Topic
+import org.oppia.android.app.viewmodel.ObservableViewModel
+
+/** [ViewModel] for displaying a topic summary. */
```
Ditto here--more detail is needed.

yashraj-01 replied:
KDoc updated. PTAL.

BenHenning reviewed (changes requested):
Thanks @yashraj-01. Took a pass on the BUILD file changes & the related view models/listeners. Mostly nits, but I'm not looking at the design or UI implementation closely.

yashraj-01 committed:
Resolved merge conflicts

yashraj-01 committed:
updated Kdocs. Nit fixes

yashraj-01 committed:
Code refactor for topics and stories

yashraj-01 committed:
Code refactored for chapters

yashraj-01 commented:
> Thanks @yashraj-01. Took a pass on the BUILD file changes & the related view models/listeners. Mostly nits, but I'm not looking at the design or UI implementation closely.

Thanks @BenHenning for the reviews. I have addressed them all. PTAL.

yashraj-01 committed:
Fixed bazel build failure

Sarthak2601 reviewed (approved):
LGTM, thanks!

BenHenning commented on app/src/main/java/org/oppia/android/app/devoptions/marktopicscompleted/MarkTopicsCompletedViewModel.kt:
```diff
@@ -61,18 +55,19 @@ class MarkTopicsCompletedViewModel @Inject constructor(
+    this.profileId = profileId
   }
+
+  /** Returns a list of [TopicViewModel] whose progress can be modified. */
```
```suggestion
  /** Returns a list of [TopicViewModel]s whose progress can be modified. */
```

Ditto elsewhere (where applicable).

yashraj-01 replied:
Done.

BenHenning reviewed (approved):
Thanks @yashraj-01. LGTM for codeowners. Just had one small nit comment--no need for me to verify it.

anandwana001 commented:
@yashraj-01 Is my approval required here, as I am not the code owner?

yashraj-01 committed:
Fixed failing tests

yashraj-01 committed:
Nit fixes

yashraj-01 commented:
> Thanks @yashraj-01. LGTM for codeowners. Just had one small nit comment--no need for me to verify it.

Thanks @BenHenning for the reviews. Also, I have made the last nit change you mentioned.

yashraj-01 commented:
> @yashraj-01 Is my approval required here, as I am not the code owner?

Ah! Actually, when I requested you for review, I didn't know about the code owner thing. Also, since you have requested changes on this PR I think you have to approve them also. Assigning you now. Thanks.

yashraj-01 committed:
Merge branch 'develop' into ui-modify-lesson-progress

rt4914 commented on app/src/main/java/org/oppia/android/app/devoptions/markchapterscompleted/MarkChaptersCompletedFragmentPresenter.kt:
```diff
@@ -0,0 +1,195 @@
+    ) {
+      bindingAdapter.notifyItemRangeChanged(
+        chapterIndex,
+        /*itemCount = */nextStoryIndex - chapterIndex
```
Nit space `/*itemCount = */ nextStoryIndex - chapterIndex`

yashraj-01 replied:
Done. Also, thanks for the review.

rt4914 reviewed (approved):
LGTM, thanks. One nit suggested.

yashraj-01 committed:
Nit fix

anandwana001 reviewed (approved):
LGTM

yashraj-01 committed:
Resolved merge conflicts

yashraj-01 committed:
Fixed failing build and tests

yashraj-01 committed:
Added files to test_file_exemptions
